src/main/java/org/folio/circulation/services/EventPublishingService.java

```diff
@@ -0,0 +1,171 @@
+import static org.folio.rest.util.OkapiConnectionParams.OKAPI_TENANT_HEADER;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.folio.circulation.domain.CheckInProcessRecords;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.domain.RequestAndRelatedRecords;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.folio.rest.jaxrs.model.Event;
+import org.folio.rest.jaxrs.model.EventMetadata;
+import org.folio.rest.util.OkapiConnectionParams;
+import org.folio.util.pubsub.PubSubClientUtils;
+
+import io.vertx.core.Context;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.web.RoutingContext;
+
+public class EventPublishingService {
+  private static final Logger logger = LoggerFactory.getLogger(EventPublishingService.class);
+
+  public static final String ITEM_CHECKED_OUT_EVENT_TYPE = "ITEM_CHECKED_OUT";
+  public static final String ITEM_CHECKED_IN_EVENT_TYPE = "ITEM_CHECKED_IN";
+  public static final String ITEM_DECLARED_LOST_EVENT_TYPE = "ITEM_DECLARED_LOST";
+  public static final String LOAN_DUE_DATE_UPDATED_EVENT_TYPE = "LOAN_DUE_DATE_UPDATED";
+
+  public static final String USER_ID_FIELD = "userId";
+  public static final String LOAN_ID_FIELD = "loanId";
+  public static final String DUE_DATE_FIELD = "dueDate";
+  public static final String RETURN_DATE_FIELD = "returnDate";
+  public static final String DUE_DATE_CHANGED_BY_RECALL_FIELD = "dueDateChangedByRecall";
+
+  private final Map<String, String> okapiHeaders;
+  private final Context vertxContext;
+
+  private Clients clients;
+
+  public EventPublishingService(RoutingContext routingContext) {
+    okapiHeaders = routingContext.request().headers().entries().stream()
+      .collect(Collectors.toMap(entry -> entry.getKey().toLowerCase(), Map.Entry::getValue));
+    vertxContext = routingContext.vertx().getOrCreateContext();
+  }
+
+  public EventPublishingService(RoutingContext routingContext, Clients circulationClients) {
+    this(routingContext);
+    clients = circulationClients;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> publishItemCheckedOutEvent(
+    LoanAndRelatedRecords loanAndRelatedRecords) {
+
+    if (loanAndRelatedRecords.getLoan() != null) {
+      Loan loan = loanAndRelatedRecords.getLoan();
+
+      JsonObject payloadJsonObject = new JsonObject();
+      write(payloadJsonObject, USER_ID_FIELD, loan.getUserId());
+      write(payloadJsonObject, LOAN_ID_FIELD, loan.getId());
+      write(payloadJsonObject, DUE_DATE_FIELD, loan.getDueDate());
+
+      publishEvent(ITEM_CHECKED_OUT_EVENT_TYPE, payloadJsonObject.encode());
```

Comment:
Does this mean that publishing the event happens completely in the background? What happens if publishing fails?

Reply:
It's ignored. Do you think we should fail the whole process (check in, check out etc.)?

Reply:
I think we don't have a general answer to that question yet. I think it's potentially a question for the POs.

In the meantime, we've typically been failing the process hard, as that at least makes the circumstances observable.

Reply:
Changed it to fail when can't publish to PubSub